src/media/inner.rs

```diff
@@ -1652,3 +1661,137 @@ fn get_or_create_source_tx<'a>(
         sources_tx.last_mut().unwrap()
     }
 }
+
+#[derive(Debug)]
+pub struct RtxCache {
+    max_packet_count: usize,
+    max_packet_age: Duration,
+    packet_by_seq_no: BTreeMap<SeqNo, Packetized>,
+    seq_no_by_quantized_size: BTreeMap<usize, SeqNo>,
+    last_sent_time: Option<Instant>,
+}
+
+const RTX_CACHE_SIZE_QUANTIZER: usize = 25;
+
+impl RtxCache {
+    fn new(max_packet_count: usize, max_packet_age: Duration) -> Self {
+        Self {
+            max_packet_count,
+            max_packet_age,
+            packet_by_seq_no: BTreeMap::new(),
+            seq_no_by_quantized_size: BTreeMap::new(),
+            last_sent_time: None,
+        }
+    }
+
+    fn cache_sent_packet(&mut self, seq_no: SeqNo, packet: Packetized, now: Instant) {
+        self.packet_by_seq_no.insert(seq_no, packet);
+        self.last_sent_time = Some(now);
+        self.remove_old_packets(now);
+    }
+
+    fn evict_acked_packet_by_seq_no(&mut self, acked_seq_no: SeqNo) {
+        self.packet_by_seq_no.remove(&acked_seq_no);
+    }
+
+    fn oldest_cached_seq_no(&self) -> Option<SeqNo> {
+        self.packet_by_seq_no.keys().next().copied()
+    }
+
+    fn last_sent_time(&self) -> Option<Instant> {
+        self.last_sent_time
+    }
+
+    fn get_cached_packet_by_seq_no(&self, seq_no: SeqNo) -> Option<&Packetized> {
+        self.packet_by_seq_no.get(&seq_no)
+    }
+
+    fn get_cached_packet_smaller_than(&self, max_size: usize) -> Option<(SeqNo, &Packetized)> {
+        let quantized_size = max_size / RTX_CACHE_SIZE_QUANTIZER;
+        let seq_no = self
+            .seq_no_by_quantized_size
+            .range(..=quantized_size)
+            .next_back()?
+            .1
+            .clone();
+        Some((seq_no, self.get_cached_packet_by_seq_no(seq_no)?))
```

Comment:
This will return None if the largest packet found has been evicted already, I think this would be better:


```rust
self
            .seq_no_by_quantized_size
            .range(..=quantized_size)
            .rev()
            .find_map(|(_, seq_no)| self.get_cached_packet_by_seq_no(seq_no))
            .cloned()

```

From our tests it seems preferable to use spurious RTX for padding than generate blank packets.